### PR TITLE
feat: migrate interactive guides API to the new Grafana App Platform group

### PR DESCRIPTION
**Pull request by tomglenn**
## What

Migrates the plugin's existing InteractiveGuide App Platform integration from the old Cloud App Platform group `pathfinderbackend.ext.grafana.com` to the new Grafana App Platform group `pathfinderbackend.ext.grafana.app`. CAP is being deprecated. This is a **hard cutover** — the plugin talks only to GAP.

## Why a hard cutover (not dual-write)

The release is sequenced so the plugin never faces a GAP backend without data: GAP is rolled out to all stacks and the GAP team's idempotent CAP→GAP backfill runs **before** the plugin release. The availability toggle already hides the feature cleanly on any stack where GAP isn't rolled out yet. So there's no transition window to bridge — an earlier dual-write/read-fallback design (see history) was over-engineered and was the source of the blocking review issues; it's been removed.

## How

- New shared `src/utils/interactive-guides-api.ts` — single source of truth for the GAP group/version/resource, `isBackendApiAvailable()` (gates on `aggregation.pathfinderbackend-ext-grafana-app.enabled`, confirmed with the GAP team), and URL builders. No dual-write, union, or fallback.
- All call sites do single-group GAP reads/writes: `fetchBackendGuides`, the block-editor guide CRUD (save/publish/unpublish/delete), the `backend-guide:` loader, and the context-init probe.
- MCP `finalize` handoff points at GAP only, with its original "GAP → grafanaOss on 404" guidance unchanged.
- Migrated the CAP-only external tooling + reference docs to `.app`: `scripts/upsert-guide.sh`, `docs/developer/EXTERNAL_API.md`, `docs/developer/CUSTOM_GUIDES.md`, `docs/sources/architecture/_index.md`, and the App Platform handoff contract.

## Scope

Frontend-only — **no `pkg/plugin` changes**. Does not touch the completion-records proxy or the deferred custom-guide-repository catalogue proxy: both remain on the old group and are documented as such in `docs/design/BACKEND_PROXY_PATTERN.md` (out of scope here, so its `.com` references stay accurate). Design/phase archive docs likewise retain their historical `.com` references.

## Laggard handling (out-of-band)

A stack that stays on the old plugin and keeps authoring against CAP after release, then upgrades, briefly won't see those guides — **no data loss** (CAP is additive, never deleted). Resolved by re-running the idempotent CAP→GAP migration sweep until CAP retirement; retirement is gated on a final reconciliation sweep. Kept in the migration script rather than the plugin (a plugin read-fallback reintroduces write/delete asymmetries).

## Verification

- New `src/utils/interactive-guides-api.test.ts` (availability matrix + URL builders); existing `backend-guide` and `finalize` tests updated / snapshot regenerated.
- typecheck · eslint · prettier · architecture/import-graph governance all pass; full suite green apart from a pre-existing Node-version test artifact unrelated to this change.
- E2E on a GAP dev stack: author + publish via the block editor → reads/writes hit the `.app` group; feature hides cleanly on a stack without the GAP toggle.

### CODE REVIEW

**Review by moxious — CHANGES_REQUESTED** [review]
Request changes. The migration direction and test coverage are solid, but there are three blocking state/contract issues to address before merge.

### [blocking] Legacy-only guides cannot be mutated before backfill — High

`readListMerged` deliberately exposes guides that exist only in the legacy group, but `dualWrite` always writes to the new group first and propagates any failure before attempting legacy.

On a dual-toggle stack before backfill, editing, publishing, unpublishing, or deleting a visible legacy-only guide sends a PUT/DELETE to the new group, receives 404, and never reaches the existing legacy resource. A partial delete can also become sticky: if the new delete succeeds and the legacy delete fails, the legacy copy reappears after refresh, while the next delete stops on the new group’s 404.

Please make item mutations group-aware:

- Allow a missing/unavailable new-group item to fall through to legacy.
- Preserve and use each group’s own `resourceVersion`.
- Ensure a missing primary copy does not prevent cleanup of the secondary copy.
- Add integration coverage for legacy-only update/publish/delete and partial-delete retry.

Avoid blindly treating every current `UNAVAILABLE_STATUSES` value as write fallback: in particular, confirm the intended authorization semantics before falling through on 403.

### [blocking] The migration is not rollback-safe — High

New-group success is authoritative while legacy failures are only logged and returned through `secondaryFailures`; every block-editor caller discards that result. The group-scoped `resourceVersion` from the new-wins merged item also makes legacy secondary PUTs likely to 409, leaving legacy stale by design. The MCP generic path similarly writes only the new group when that write succeeds.

A code revert restores the old-only client, which can then:

- Lose guides created only in GAP.
- Show stale versions after CAP secondary-write failures.
- Resurrect resources deleted only from GAP.

Please define and verify the rollback/reconciliation procedure before merge: feature-toggle ordering, how GAP-authoritative creates and updates are replayed to CAP, how deletions are represented, and how divergence is detected and repaired. Ideally, make secondary divergence durable/actionable rather than returning a result callers ignore.

### [blocking] The MCP handoff gives contradictory 404 instructions — High

The handoff currently specifies mutually exclusive behavior for the same collection 404:

- `appPlatform.errorHandling` and `grafanaAppPlatform.appliesWhen` say to retry the legacy group.
- The detailed error table says to switch immediately to `grafanaOss` and “Do not retry.”
- `grafanaOss.appliesWhen` and the top-level instructions also permit immediate OSS fallback.
- Preferred-tool guidance mentions a legacy retry, but the generic-write fallback is restricted to environments where preferred tools are not registered.

Depending on which field an Assistant follows, an old-only stack either receives the intended legacy retry or incorrectly falls through to local export.

Please define one deterministic new → legacy → OSS state machine, apply it consistently across every handoff field, update the App Platform handoff contract, and add focused semantic assertions rather than relying only on the broad snapshot.

### Other findings

- **Medium — External import tooling remains CAP-only.** `scripts/upsert-guide.sh`, `docs/developer/EXTERNAL_API.md`, and `docs/developer/CUSTOM_GUIDES.md` still hard-code the deprecated group. They continue creating legacy-only resources during migration and will fail after CAP retirement. Please migrate them or explicitly track that work before retirement. A Node-safe core module for the shared group/version/resource constants would also prevent the browser and MCP definitions from drifting.
- **Low — One group’s transient failure blanks the merged list.** `readListMerged` rethrows a genuine error from either group and discards data already available from the other. On a dual stack, consider returning the successful group’s data and recording a degradation, throwing only when no active group succeeds.
- **Instrumentation — Non-blocking.** Item fallback, list-group skipping, and failed secondary writes are degradation paths. `TELEMETRY.md` requires a typed, low-cardinality facade event so these paths are countable and alertable; the logger warning alone does not cover that requirement.
- **Security question — Non-blocking.** A high-severity authorization-bypass claim was not substantiated because the legacy endpoint independently authorizes the caller. Still, please confirm that RBAC is equivalent across groups before retaining 403 as an “unavailable, continue” result.
- **Tech debt — Non-blocking.** The changed guide API code would benefit from a shared typed `InteractiveGuideResource`/metadata contract instead of the current `any` clusters.

Verification: GitHub reports 35 passed checks and 0 failures. Focused local verification also passed: 39 tests, typecheck, and lint with zero errors. The blocking findings above concern migration semantics that the current tests either miss or explicitly encode.

**Comment by tomglenn** [comment]
Thanks David — this pushed us to reconsider the whole approach, and you were right that the dual-write was carrying its weight in bugs. We've **dropped dual-write entirely and reworked this to a hard cutover** (pushed in `refactor: simplify GAP migration to a hard cutover` + `docs: migrate external guide tooling`).

The reasoning: the plugin release is gated on GAP being rolled out everywhere **and** the idempotent CAP→GAP backfill running **first**, so the plugin never faces a GAP backend that lacks the data. The availability toggle already hides the feature cleanly on not-yet-GAP stacks. That removes the transition window your three blocking issues all stem from:

- **[Blocking] Legacy-only mutation** — gone. There are no legacy-only guides once backfill runs before release; the plugin only ever talks to GAP.
- **[Blocking] Rollback-safety** — collapses to a normal, documented cutover risk (CAP is additive/never deleted; retirement gated on a final reconciliation sweep), rather than a dual-write that — as you noted — didn't actually keep CAP in sync anyway.
- **[Blocking] Contradictory MCP 404 guidance** — gone. The handoff reverts to the single clean "GAP → grafanaOss" state machine, just pointed at `.app`; the legacy-retry branch that conflicted with the existing guidance is removed.

The one residual we accept explicitly: a laggard stack that keeps authoring on CAP after release, then upgrades before the next sweep, briefly won't see those guides — no data loss, resolved on the next idempotent sweep. Handling that in the plugin (read-fallback) reintroduced exactly the write/delete asymmetries you flagged, so we've deliberately kept it in the migration script instead.

On the **non-blocking** items:
- **External tooling / drift** — migrated `upsert-guide.sh`, `EXTERNAL_API.md`, `CUSTOM_GUIDES.md`, the architecture doc, and the handoff contract to `.app`; group/version/resource now live in one shared `src/utils/interactive-guides-api.ts` (design/phase archives keep their historical `.com`).
- **List-blanking on transient error / read-fallback telemetry** — no longer applicable (no union/fallback reads).
- **403 auth semantics** — no longer relevant (no cross-group write fallback).
- **Typed `InteractiveGuideResource` contract** — fair; tracking as follow-up tech-debt, out of scope for this now-smaller change.

Net effect is a much smaller diff: a `.com`→`.app` cutover behind the availability toggle. Ready for another look when you have a moment.

**Comment by moxious** [comment]
Filed two follow-ups from the review, to be handled outside this PR:

- #1431 — migrate the remaining design docs (`AI-AUTHORING-IMPLEMENTATION.md`, `BACKEND_PROXY_PATTERN.md`, two `phases/ai-authoring-*`) off the legacy CAP group (low-priority doc drift).
- #1432 — research + align the Grafana Assistant authoring tools (`pathfinder_manage_guide_drafts` / `pathfinder_publish_guide`, implemented cross-repo in Assistant) with the new GAP group. This PR migrated the in-repo `finalize` handoff + generic-write fallback, but the Assistant preferred-tool write path is out of scope here and needs cross-repo verification.

**Review by moxious — CHANGES_REQUESTED** [review]
Re-review:  Filing follow-up issues https://github.com/grafana/grafana-pathfinder-app/issues/1432 and https://github.com/grafana/grafana-pathfinder-app/issues/1431

PR needs new description but is otherwise ok

**Review by moxious — DISMISSED** [review]

File: src/utils/interactive-guides-api.ts [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+      secondaryFailures.push(group);
+      logger.warn('[interactive-guides-api] secondary dual-write failed', {
+        group,
+        method: args.method,
+        status: extractStatus(err),
+      });
+    }
+  }
+  return { primaryGroup: primary, secondaryFailures };
+}
```
[the remark this answers is not in the document]
Comment (tomglenn) [reply src/utils/interactive-guides-api.ts:27]: Resolved by the rework — the dual-write path is gone. This PR is now a hard cutover: the plugin talks only to GAP, so there is no primary/secondary write and no legacy fallback. `dualWrite` no longer exists; writes are single-group GAP calls in `useBackendGuides.ts`. The legacy-only-mutation race this describes can no longer occur (there are no legacy-only guides once the GAP backfill runs before release — see the updated PR description).

File: src/cli/mcp/tools/finalize.ts [hunk 2]
```diff
@@ -164,7 +173,7 @@ async function finalizeImpl(args: {
     clientGuidance: {
       grafanaAppPlatform: {
         appliesWhen:
-          'You are a Grafana-aware client (e.g., Grafana Assistant) running inside a Grafana instance that has the Pathfinder backend aggregator enabled (featureToggles["aggregation.pathfinderbackend-ext-grafana-com.enabled"] === true). If you are unsure, try this branch first; on a 404 from the collection POST or from pathfinder_manage_guide_drafts, switch to grafanaOss.',
+          'You are a Grafana-aware client (e.g., Grafana Assistant) running inside a Grafana instance that has the Pathfinder backend aggregator enabled (featureToggles["aggregation.pathfinderbackend-ext-grafana-app.enabled"] === true, or the legacy featureToggles["aggregation.pathfinderbackend-ext-grafana-com.enabled"] during the CAP→GAP migration). If you are unsure, try this branch first; on a 404 from the collection POST or from pathfinder_manage_guide_drafts, retry against appPlatform.legacyApiVersion, then switch to grafanaOss.',
```
[the remark this answers is not in the document]
Comment (tomglenn) [reply src/cli/mcp/tools/finalize.ts]: Resolved by the rework — the legacy-retry guidance was removed. `finalize.ts` now points at GAP only and keeps its original single "GAP → grafanaOss on 404" state machine, so there is no longer any conflicting 404 instruction across the handoff fields.

**Review by jdbaldry — APPROVED** [review]
Trivial change since David's previous review LGTM
